### ISSUE TITLE
Add ``.install plugin url`` and ``.ip url``Command (Would Love To Do Myself but Too Chaotic Yet, Will Keep Looking Into It)

### ISSUE THREAD

**Issue by Minionguyjpro** [Osiris-Team/AutoPlug-Client#237]
**Importance:** 5/10

**Problem:**
The functionality to download a plugin using a direct download link/URL to the .jar is missing. This functionality would allow us to install plugins easily that are outside of the known platforms such as Bukkit, Spigot and Modrinth (GitHub too, but that's for open source projects in general).

**Solution:**
Add a command to install a plugin using an URL directly, and throw an error if the downloading file from the URL doesn't end in ``.jar``.

**Comment by Osiris-Team** [comment]
@Minionguyjpro Yeah sounds good, I would even extend it to make installing possible from, for example a spigot url, which would then extract the type (spigot) and id from the url, and then execute `.install plugin spigot 1234` internally, and thus make the whole process much more convenient.

**Comment by Minionguyjpro** [comment]
I'm going to make another PR for the client soon to migrate it to Gradle, though it could conflict with the already open PRs. They'd have to make sure to sync the changes if it gets merged!

**Comment by Osiris-Team** [comment]
@Minionguyjpro sure give that a try, it could be a bit harder though since it depends on some plugins.

**Comment by Minionguyjpro** [comment]
> @Minionguyjpro sure give that a try, it could be a bit harder though since it depends on some maven plugins.

True, but we use maven central as repository when building with Gradle! So I guess this won't be a problem at all.

**Comment by Osiris-Team** [comment]
@Minionguyjpro as far as I know maven plugins don't run on Gradle. You will have to find alternative Gradle plugins that do the same things.

**Comment by Minionguyjpro** [comment]
No, there is no need for that actually since Gradle supports the Maven Central repository and it can download all of the same plugins from there just fine and build (I tested). I'm not sure though, what is the difference between the generated ``AutoPlug-Client.jar`` in ``AP-TEST-SERVER/`` compared to the ``original-AutoPlug-Client.jar`` in ``target/``. The one in the ``target``directory seems to be bigger. Please explain to me, it can help with some migration stuff to Gradle.

**Comment by Minionguyjpro** [comment]
> No, there is no need for that actually since Gradle supports the Maven Central repository and it can download all of the same plugins from there just fine and build (I tested). I'm not sure though, what is the difference between the generated `AutoPlug-Client.jar` in `AP-TEST-SERVER/` compared to the `original-AutoPlug-Client.jar` in `target/`. The one in the `target`directory seems to be bigger. Please explain to me, it can help with some migration stuff to Gradle.

Well maybe they aren't runnable on Gradle. But maybe the ones in the maven Central also have a Gradle version? I'm not sure, but all plugins for Minecraft projects that ran on Maven, also seem to have a version that can be used by Gradle.

**Comment by Minionguyjpro** [comment]
Oh it generates them, but could you just explain me what the AP-TEST-SERVER
one is different from the other generated one in target?

Op di 30 apr. 2024 22:10 schreef Osiris-Team ***@***.***>:

**Comment by Osiris-Team** [comment]
@Minionguyjpro Ahh thats what you meant. The jar in AP-TEST-SERVER is simply the shaded jar that was moved from /target into that directory. 

In AP-TEST-SERVER, you can run autoplug and do manual tests.

**Comment by Minionguyjpro** [comment]
> @Minionguyjpro Ahh thats what you meant. The jar in AP-TEST-SERVER is simply the shaded jar that was moved from /target into that directory.
> 
> In AP-TEST-SERVER, you can then run autoplug and do manual tests.

I see. But is the jar in /target not shaded or is it? I want to try to replicate the functionality as close as possible.

**Comment by Minionguyjpro** [comment]
> > @Minionguyjpro Ahh thats what you meant. The jar in AP-TEST-SERVER is simply the shaded jar that was moved from /target into that directory.
> > In AP-TEST-SERVER, you can then run autoplug and do manual tests.
> 
> I see. But is the jar in /target not shaded or is it? I want to try to replicate the functionality as close as possible.

I see it's not, okay.

**Comment by Minionguyjpro** [comment]
![afbeelding](https://github.com/Osiris-Team/AutoPlug-Client/assets/66115754/5a3ddaea-ee69-4cbf-910e-d60164860372)
Okay @Osiris-Team I got tests to work, but do you know why this one here fails?

**Comment by Minionguyjpro** [comment]
> ![afbeelding](https://private-user-images.githubusercontent.com/66115754/327094796-5a3ddaea-ee69-4cbf-910e-d60164860372.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1NTY1MjIsIm5iZiI6MTcxNDU1NjIyMiwicGF0aCI6Ii82NjExNTc1NC8zMjcwOTQ3OTYtNWEzZGRhZWEtZWU2OS00Y2JmLTkxMGUtZDYwMTY0ODYwMzcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDA5MzcwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhMzlkNmZlYjIzYjgzZGE5YWEyY2Q4MDNiYzk5ODg4NGFlMTlmNDY3ZWJkN2ViMWQyNjBjZmFlZDY4ZGVjZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QhUtRhrnwqdjvjPeJZIkgA_nY7eHHYbrpmhtwR470vM) Okay @Osiris-Team I got tests to work, but do you know why this one here fails?

Maven also seems to fail the same test.

**Comment by Minionguyjpro** [comment]
Ah nevermind you exclude the test task. I now successfully got to replicate the same names (Shaded ``AutoPlug-Client.jar`` in ``AP-TEST-SERVER`` and ``original-AutoPlug-Client.jar`` in ``build/libs``). Will need to do some manual tests myself to see whether it runs.

**Comment by Osiris-Team** [comment]
In pom.xml there is a section that creates and includes an autoplug.properties file in the jar, so you must replicate that too in gradle.

**Comment by Minionguyjpro** [comment]
> @Minionguyjpro in pom.xml there is a section that creates and includes an autoplug.properties file in the jar with the version, so you must replicate that too in gradle.

Okay got it, was able to recreate such a thing in Gradle using writing and stuff. It works, the version is now visible but I still get this error on startup after some time:
```
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.
SLF4J: Ignoring binding found at [jar:file:/home/minionguyjpro/Documenten/GitHub/AutoPlug-Client/AP-TEST-SERVER/AutoPlug-Client.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See https://www.slf4j.org/codes.html#ignoredBindings for an explanation
```

**Comment by Minionguyjpro** [comment]
> > @Minionguyjpro in pom.xml there is a section that creates and includes an autoplug.properties file in the jar with the version, so you must replicate that too in gradle.
> 
> Okay got it, was able to recreate such a thing in Gradle using writing and stuff. It works, the version is now visible but I still get this error on startup after some time:
> 
> ```
> SLF4J: No SLF4J providers were found.
> SLF4J: Defaulting to no-operation (NOP) logger implementation
> SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
> SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.
> SLF4J: Ignoring binding found at [jar:file:/home/minionguyjpro/Documenten/GitHub/AutoPlug-Client/AP-TEST-SERVER/AutoPlug-Client.jar!/org/slf4j/impl/StaticLoggerBinder.class]
> SLF4J: See https://www.slf4j.org/codes.html#ignoredBindings for an explanation
> ```

And it seems like AutoPlug functions as normal apart from that! I want to get that error fixed so we can get that annoying thing away.

**Comment by Osiris-Team** [comment]
Also handled in the pom.xml somewhere. Search for slf4j in there.
